alg/teca_temporal_reduction.cxx

```diff
@@ -0,0 +1,2324 @@
+#include "teca_temporal_reduction.h"
+
+#include "teca_cartesian_mesh.h"
+#include "teca_array_collection.h"
+#include "teca_variant_array.h"
+#include "teca_variant_array_impl.h"
+#include "teca_metadata.h"
+#include "teca_calendar_util.h"
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <cmath>
+
+#if defined(TECA_HAS_BOOST)
+#include <boost/program_options.hpp>
+#endif
+
+#if defined(TECA_HAS_CUDA)
+#include "teca_cuda_util.h"
+#endif
+
+//#define TECA_DEBUG
+
+using std::string;
+using std::vector;
+using std::cerr;
+using std::endl;
+using std::cos;
+
+// PIMPL idiom hides internals
+// defines the API for reduction operators
+class teca_temporal_reduction::internals_t
+{
+public:
+    internals_t() {}
+    ~internals_t() {}
+
+public:
+    class reduction_operator;
+    class average_operator;
+    class summation_operator;
+    class minimum_operator;
+    class maximum_operator;
+    class reduction_operator_factory;
+    class time_interval;
+
+    using p_reduction_operator = std::shared_ptr<reduction_operator>;
+
+public:
+    std::vector<time_interval> indices;
+    std::map<std::string, p_reduction_operator> op;
+};
+
+class teca_temporal_reduction::internals_t::reduction_operator
+{
+public:
+    /** reduction_operator implements 4 class methods:
+     *                    initialize, update_cpu, update_gpu, and finalize.
+     *
+     *  initialize:
+     *      initializes the reduction. If a fill value is passed
+     *      the operator should use it to identify missing values in the
+     *      data and handle them appropriately.
+     *  update:
+     *      reduces arrays_in (new data) into arrays_out (current state) and
+     *      returns the result. For update_cpu, the target device is cpu;
+     *      and for update_gpu, the target device is gpu (cuda).
+     *  finalize:
+     *      finalizes arrays_out (current state) and returns the result.
+     *      If no finalization is needed simply return arrays_out.
+     */
+
+    reduction_operator() : fill_value(-1) {}
+
+    virtual void initialize(double fill_value);
+
+    virtual int update_cpu(int device_id,
```

Comment:
there are a bunch of compiler warnings relating to these functions, eg not return statement.  I think you could make these pure virtual since it would be a problem if the derived classes did not implement them.